topics/dev/tutorials/core-contributing/tutorial.md

```diff
@@ -0,0 +1,219 @@
+
+Please review ``test_users.py`` and attempt to write a test case that:
+
+- Verifies the test user's initially favorited extensions is an empty list.
+- Verifies that a ``POST`` to ``<galaxy_root_url>/api/users/current/favorites/extensions/fasta`` returns a 200 status code indicating success.
+- Verifies that after this ``POST`` the list of user favorited extensions contains ``fasta`` and is of size 1.
+- Verifies that a ``DELETE`` to ``<galaxy_root_url>/api/users/current/favorites/extensions/fasta`` succeeds.
+- Verifies that after this ``DELETE`` the favorited extensions list is again empty.
+
+{% include topics/dev/tutorials/core-contributing/test_users.py_diff.md %}
+
+Verify this test fails when running stand-alone.
+
+```
+./run_tests.sh -api lib/galaxy_test/api/test_users.py::UsersApiTestCase::test_favorite_extensions
+```
+
+Add a new API implementation file to
+``lib/galaxy/webapps/galaxy/api/`` called ``user_favorites.py`` with
+an API implementation of the endpoints we just outlined.
+
+To implement the API itself, add three methods to the user manager in
+``lib/galaxy/managers/users.py``.
+
+- ``get_favorite_extensions(user)``
+- ``add_favorite_extension(user, extension)``
+- ``delete_favorite_extension(user, extension)``
+
+{% include topics/dev/tutorials/core-contributing/user_favorites.py_diff.md %}
+
+{% include topics/dev/tutorials/core-contributing/users.py_diff.md %}
+
+This part is relatively challenging and takes time to really become an
+expert at - it requires greping around the backend to find similar examples,
+lots of trial and error, debugging the test case and the implementation in unison,
+etc..
+
+Ideally, you'd start at the top of test case - make sure it fails on the first API request,
+implement ``get_favorite_extensions`` on the manager and the API code to wire it up, and continue
+with ``add_favorite_extension`` before finishing with ``delete_favorite_extension``.
+
+
+```
+./run_tests.sh -api lib/galaxy_test/api/test_users.py::UsersApiTestCase::test_favorite_extensions
+```
+
+Once the API test is done, it is time to build a user interface for this addition to Galaxy. Lets get
+some of the plumbing out of the way right away. We'd like to have a URL for viewing the current user's
+favorite extensions in the UI. This URL needs to be registered as a client route in ``lib/galaxy/webapps/galaxy/buildapp.py``.
+
+Add ``/user/favorite/extensions`` as a route for the client in ``buildapp.py.
+
+{% include topics/dev/tutorials/core-contributing/buildapp.py_diff.md %}
+
+Let's add the ability to navigate to this URL and future component to
+the "User" menu in the Galaxy masthead. The file ``client/src/layout/menu.js``
+contains the "model" data describing the masthead. Add a link to the route
+we described above to ``menu.js`` with a entry titled "Favorite Extensions".
+
+{% include topics/dev/tutorials/core-contributing/menu.js_diff.md %}
+
+The next piece of this plumbing is to respond to this route the
```

Comment:
something is missing here. Respond **in** the analysis router?

Reply:
Yes - you're right - thanks for the detailed review!